log/log.go

```diff
@@ -3,142 +3,129 @@
 package log
 
 import (
-	"fmt"
+	"go.uber.org/zap/zapcore"
+	"gopkg.in/natefinch/lumberjack.v2"
+	"io"
 	"os"
 	"path/filepath"
 
-	"gopkg.in/natefinch/lumberjack.v2"
-	"gopkg.in/op/go-logging.v1"
+	"go.uber.org/zap"
 )
 
-// Log is an exported type that embeds our logger.
-// logging library can be replaced as long as it implements same functionality used across the project.
-type Log struct {
-	*logging.Logger
-}
 
-// smlogger is the local app singleton logger.
-var AppLog Log
-var debugMode = false
+const mainLoggerName = "Spacemesh"
 
-func init() {
-
-	// create a basic temp os.Stdout logger
-	// This logger is used until the app calls InitSpacemeshLoggingSystem().
-	log := logging.MustGetLogger("app")
-	log.ExtraCalldepth = 1
-	logFormat := ` %{color}%{level:.4s} %{id:03x} %{time:15:04:05.000} ▶%{color:reset} %{message}`
-	leveledBackend := getBackendLevel("app", "<APP>", logFormat)
-	log.SetBackend(leveledBackend)
-	AppLog = Log{Logger: log}
-}
+// determine the level of messages we show.
+var debugMode = false
+// should we format out logs in json
+var jsonLog = false
 
-// getAllBackend returns level backends with an exception to Debug leve
-// which is only returned if the flag debug is set
-func getBackendLevel(module, prefix, format string) logging.LeveledBackend {
-	logFormat := logging.MustStringFormatter(format)
+var DebugLevel = zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+	return lvl >= zapcore.DebugLevel
+})
 
-	backend := logging.NewLogBackend(os.Stdout, prefix, 0)
-	backendFormatter := logging.NewBackendFormatter(backend, logFormat)
-	leveledBackend := logging.AddModuleLevel(backendFormatter)
+var InfoLevel = zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+	return lvl >= zapcore.InfoLevel
+})
 
+func logLevel() zap.LevelEnablerFunc {
 	if debugMode {
-		leveledBackend.SetLevel(logging.DEBUG, module)
+		return DebugLevel
 	} else {
-		leveledBackend.SetLevel(logging.INFO, module)
+		return InfoLevel
 	}
+}
+
+func encoder() zapcore.Encoder {
+	if jsonLog || debugMode {
```

Comment:
debug will also be formatted to json?

Reply:
yes, I though when we're testing with json we'd like max verbosity. but debugging not necessarily means we want json. I'll split them.

Reply:
fixed